api/internal/handler/data_loader/route_import.go

```diff
@@ -19,535 +19,317 @@ package data_loader
-	})
-	if err != nil {
-		return err
-	}
-	if len(ret.Rows) > 0 {
-		return consts.InvalidParam("route is duplicate")
-	}
-	return nil
-}
-
-func parseExtension(val *openapi3.Operation) (*entity.Route, error) {
-	routeMap := map[string]interface{}{}
-	for key, val := range val.Extensions {
-		if strings.HasPrefix(key, "x-apisix-") {
-			routeMap[strings.TrimPrefix(key, "x-apisix-")] = val
 		}
-	}
-
-	route := new(entity.Route)
-	routeJson, err := json.Marshal(routeMap)
-	if err != nil {
-		return nil, err
-	}
-
-	err = json.Unmarshal(routeJson, &route)
-	if err != nil {
-		return nil, err
-	}
-
-	return route, nil
-}
 
-type PathValue struct {
-	Method string
-	Value  *openapi3.Operation
-}
-
-func mergePathValue(key string, values []PathValue, swagger *openapi3.Swagger) (map[string]*entity.Route, error) {
-	var parsed []PathValue
-	var routes = map[string]*entity.Route{}
-	for _, value := range values {
-		value.Value.OperationID = strings.Replace(value.Value.OperationID, value.Method, "", 1)
-		var eq = false
-		for _, v := range parsed {
-			if utils.ValueEqual(v.Value, value.Value) {
-				eq = true
-				if routes[v.Method].Methods == nil {
-					routes[v.Method].Methods = []string{}
+		// Duplicate routes found
+		if o.TotalSize > 0 {
+			for _, row := range o.Rows {
+				r, ok := row.(*entity.Route)
+				if ok {
+					errs[store.HubKeyRoute] = append(errs[store.HubKeyRoute],
+						errors.Errorf("%s is duplicated with route %s",
+							route.Uris[0],
+							r.Name).
+							Error())
 				}
-				routes[v.Method].Methods = append(routes[v.Method].Methods, value.Method)
-			}
-		}
-		// not equal to the previous ones
-		if !eq {
-			route, err := getRouteFromPaths(value.Method, key, value.Value, swagger)
-			if err != nil {
-				return nil, err
 			}
-			routes[value.Method] = route
-			parsed = append(parsed, value)
 		}
 	}
 
-	return routes, nil
+	return errs
 }
 
-func OpenAPI3ToRoute(swagger *openapi3.Swagger) ([]*entity.Route, error) {
-	var routes []*entity.Route
-	paths := swagger.Paths
-	var upstream *entity.UpstreamDef
-	var err error
-	for k, v := range paths {
-		k = regPathRepeat.ReplaceAllString(k, "")
-		upstream = &entity.UpstreamDef{}
-		if up, ok := v.Extensions["x-apisix-upstream"]; ok {
-			err = json.Unmarshal(up.(json.RawMessage), upstream)
-			if err != nil {
-				return nil, err
-			}
-		}
+// Create parsed resources
+func (h *ImportHandler) createEntities(ctx context.Context, data *loader.DataSets) map[store.HubKey][]string {
+	errs := make(map[store.HubKey][]string)
 
-		var values []PathValue
-		if v.Get != nil {
-			value := PathValue{
-				Method: http.MethodGet,
-				Value:  v.Get,
-			}
-			values = append(values, value)
-		}
-		if v.Post != nil {
-			value := PathValue{
-				Method: http.MethodPost,
-				Value:  v.Post,
+	for _, route := range data.Routes {
+		_, err := h.routeStore.Create(ctx, &route)
+		if err != nil {
+			if errs[store.HubKeyRoute] == nil {
+				errs[store.HubKeyRoute] = []string{}
```

Comment:
The explicit initialization is not necessary because append will do it .

Reply:
Oh, GET, learned something new. I will modify it. 😳

Reply:
explicitly initialize are removed